test/balancer_test.go

```diff
@@ -1004,3 +1005,101 @@ func (s) TestMetadataInPickResult(t *testing.T) {
 		t.Fatalf("Mismatch in custom metadata received at test backend, got: %v, want %v", gotMDVal, wantMDVal)
 	}
 }
+
+// producerTestBalancerBuilder and producerTestBalancer start a producer which
+// makes an RPC before the SubConn is READY, then connects the subconn, and
+// pushes the resulting error (expected to be nil) to rpcErrChan.
+type producerTestBalancerBuilder struct {
+	rpcErrChan chan error
+}
+
+func (bb *producerTestBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	return &producerTestBalancer{cc: cc, rpcErrChan: bb.rpcErrChan}
+}
+
+const producerTestBalancerName = "producer_test_balancer"
+
+func (bb *producerTestBalancerBuilder) Name() string { return producerTestBalancerName }
+
+type producerTestBalancer struct {
+	cc         balancer.ClientConn
+	rpcErrChan chan error
+}
+
+func (b *producerTestBalancer) UpdateClientConnState(ccs balancer.ClientConnState) error {
+	// Create the subconn, but don't connect it.
+	sc, err := b.cc.NewSubConn(ccs.ResolverState.Addresses, balancer.NewSubConnOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating subconn: %v", err)
+	}
+
+	// Create the producer.  This will call the producer builder's Build
+	// method, which will try to start an RPC in a goroutine.
+	p := &testProducerBuilder{start: grpcsync.NewEvent(), rpcErrChan: b.rpcErrChan}
+	sc.GetOrBuildProducer(p)
+
+	// Wait here until the producer is about to perform the RPC, which should
+	// block until connected.
+	<-p.start.Done()
+
```

Comment:
I feel like this right here is the crux of the test. Can you add some sort of assertion that EmptyCall hasn't gone through since it's blocked, and thus doesn't return ErrSubConnNotReady. Right now, this is testing that the EmptyCall() operation on line 1074 happens fine with no error, but we're not guaranteed the scheduler doesn't run sc.Connect() before EmptyCall(), which in that ordering of operations ran, doesn't test the behavior you added/changed, since on master if the scheduler runs sc.Connect() first before EmptyCall() on line 1074, it would work (because the SubConn would be ready, then EmptyCall() -> NewStream() reads READY).

Reply:
It's possible that this happens later, but it's vastly more likely that we'll hit the line that starts the RPC before all of the mechanisms that need to happen to connect the subconn happen first.  (I.e. it switches to connecting, does network operations, etc.)

Added a select from the channel here, since it's harmless, but I don't think it will add much certainty to the test since it's still reliant on ordering.

Reply:
I wasn't shooting for exact certainty with this comment. However, now we can be sure your behavior WAI of when the SubConn isn't connected yet, no ErrSubConnAvailable gets plumbed from the call like master did if EmptyCall() does happen first (however unlikely). I get that there's still a race between the operations Connect() and EmptyCall(), but the EmptyCall() proceeding as intended and not error verifies both a. if emptycall happens first, it blocks until Connect()'s underlying operations complete and doesn't error out (your added Select), or if Connect() happens first, that's also fine and EmptyCall() should work as normal (which would also happen to work on master, but that's fine - since we did add a new check for the behavior change which will be tested in the opposite order of operations).